### ISSUE TITLE
Multi-GPUs training actually slower than single-GPU training on FF network

### ISSUE THREAD

**Issue by etienneFontaine35** [Microsoft/CNTK#3325]
Hi folks,
I'm getting used to CNTK by implementing a simple feed forward network with 3 layers of 1024 nodes on MNIST dataset.
I get pretty good results on testing, however the network is slower to train with multiple GPUs than with a single one. I have 2 GTX 1080 Ti running on my computer.
It seems the process is duplicated on each GPU and not shared amongst the GPUs.
I set the MinibatchSize to 1000, it runs over 20 epochs
The results :
    - for single GPU training --> testing error = 2.53%, training duration = 4.72s
    - for 2 GPUs training --> testing error = 2.13%, training duration = 11s

here are the command lines I used :
    "python3.5 single-training_DNN.py" for single GPU
    "mpirun --npernode 2 python3.5 multi-training_DNN.py" for multi GPU

I followed the examples from https://docs.microsoft.com/en-us/cognitive-toolkit/Multiple-GPUs-and-machines and https://tsmatz.wordpress.com/2018/01/25/cntk_distributed_training_with_multiple_machines/
if you need the source code I can also add it.
thanks,
Etienne

**Comment by Tixxx** [comment]
Hi @etienneFontaine35, since CNTK is using data-parallel training process, when multiple GPUs are involved, there might be some overhead of synchronizing parameters after each minibatch. The overhead will be dependent on the parameter size. 
Is your code one of the examples in the links you pasted? If not could you share it with us?
In the meantime, could you also try different larger MinibatchSizes and see if it improves?

**Comment by etienneFontaine35** [comment]
Hi @Tixxx , thank you for your reply !
I put my code as well as some more results :
`# Import the relevant modules to be used later
from __future__ import print_function
import gzip
import matplotlib.image as mpimg
import matplotlib.pyplot as plt
import numpy as np
import os
import shutil
import struct
import sys
import time as t
import datetime

import cntk as cntk
from cntk.train.training_session import *
from cntk.device import cpu, gpu


cntk.cntk_py.set_fixed_random_seed(42) # fix a random seed for CNTK components
# cntk.cntk_py.set_fixed_random_seed(int(t.time()))

# Read a CTF formatted text (as mentioned above) using the CTF deserializer from a file
def create_reader(path, is_training, input_dim, num_label_classes, multi_threaded):
    return cntk.io.MinibatchSource(
        cntk.io.CTFDeserializer(
            path, cntk.io.StreamDefs(
                labels = cntk.io.StreamDef(field='labels', shape=num_label_classes, is_sparse=False),
                features = cntk.io.StreamDef(field='features', shape=input_dim, is_sparse=False)
            )
        ),
        randomize = is_training,
        max_sweeps=  cntk.io.INFINITELY_REPEAT if is_training else 1,
        multithreaded_deserializer=multi_threaded # parallele
    )


def create_model(features, num_hidden_layers, hidden_layers_dim, num_output_classes):
    reseau = features
    with cntk.layers.default_options(init = cntk.layers.glorot_uniform(), activation = cntk.ops.relu):
        for _ in range(num_hidden_layers):
            reseau = cntk.layers.Dense(hidden_layers_dim)(reseau)
        reseau = cntk.layers.Dense(num_output_classes, activation=None)(reseau)
        return reseau

#####################################################################################################################
###---------------------------------------------------------------------------------------------------------------###
#####################################################################################################################

# On vérifie l'existence des données
train_file = "DataSets/Train-28x28_cntk_text.txt"
test_file = "DataSets/Test-28x28_cntk_text.txt"

if not os.path.exists(train_file) or not os.path.exists(test_file) :
	print("Fichier de Train ou de Test inexistant")
	sys.exit()
	

# Initialisation des paramètres
inputDimension = 28 * 28
numOutputClasses = 10

numHiddenLayers = int(sys.argv[1])
hiddenLayerDim = int(sys.argv[2])

learningRate = 0.2
trainMinibatchSize = 64
trainSize = 60000
numEpochs = 20
numSamplesTrain = trainSize * numEpochs

testMinibatchSize = 512
testSize = 10000
numMbTest = testSize // testMinibatchSize

trainProgressFreq = trainSize

if cntk.Communicator.rank() == 0 :
    print("CNTK version: %s"%(cntk.__version__), "\n")
    print(str(datetime.datetime.now()), "\n")

    print("Input dimension :", inputDimension)
    print("Output dimension :", numOutputClasses)
    print("Nbre layers :", numHiddenLayers)
    print("Dimension layers :", hiddenLayerDim)
    print("Nbre epoques :", numEpochs)
    print("Minibatch size :", trainMinibatchSize, "\n")

#----------------------
#   Reading Data
#----------------------

train_reader = create_reader(train_file, True, inputDimension, numOutputClasses, False)
test_reader = create_reader(test_file, False, inputDimension, numOutputClasses, False)


#----------------------
#       Modèle
#----------------------

featureImage = cntk.input_variable(inputDimension)
labelDigit = cntk.input_variable(numOutputClasses)

FFreseau = create_model(cntk.ops.element_times(cntk.ops.constant(1/255.0), featureImage), numHiddenLayers, hiddenLayerDim, numOutputClasses)

lossFunc = cntk.cross_entropy_with_softmax(FFreseau, labelDigit)
errorFunc = cntk.classification_error(FFreseau, labelDigit)



#----------------------
#      Training
#----------------------

inputMapTrain = {
	featureImage : train_reader.streams.features,
	labelDigit : train_reader.streams.labels
}

trainProgress = cntk.logging.ProgressPrinter(
    tag='Training',
    num_epochs=numEpochs,
    freq=trainProgressFreq,
    log_to_file=None,
    rank=cntk.train.distributed.Communicator.rank(), # parallele
    distributed_freq=None)  # parallele


lrSchedule = cntk.learning_rate_schedule(learningRate, cntk.UnitType.minibatch)
learner = cntk.sgd(FFreseau.parameters, lrSchedule)
dist_learner = cntk.train.distributed.data_parallel_distributed_learner( # parallele
    learner = learner, # parallele
    num_quantization_bits=1, # non-quantized SGD # parallele
    distributed_after=0) # all data is parallelized  # parallele

trainer = cntk.Trainer(FFreseau, (lossFunc, errorFunc), dist_learner, trainProgress)  # parallele : changement du learner

trainingSession = cntk.train.training_session(
			trainer=trainer,
			mb_source=train_reader,
			mb_size=trainMinibatchSize,
			model_inputs_to_streams=inputMapTrain,
			max_samples=numSamplesTrain,
			# cv_config=CrossValidationConfig(train_reader, minibatch_size=trainMinibatchSize, frequency=(2, cntk.DataUnit.sweep), max_samples=numMbTrain),
			progress_frequency=trainProgressFreq)

debut = t.time()
trainingSession.train()
fin = t.time()
print("Learning duration :", round(fin - debut, 2), "s")

if cntk.Communicator.rank() == 0 :
    FFreseau.save("models/multi_" + str(numHiddenLayers) + "-" + str(hiddenLayerDim) + "_" + str(numEpochs) + "-" + str(trainMinibatchSize) + ".dnn")


#----------------------
#      Testing
#----------------------
if cntk.Communicator.rank() == 0 : 
    inputMapTest = {
        featureImage : test_reader.streams.features,
        labelDigit : test_reader.streams.labels
    }

    testResult = 0.0

    for i in range(numMbTest) :
        mb = test_reader.next_minibatch(testMinibatchSize, input_map=inputMapTest)
        evalError = trainer.test_minibatch(mb)
        testResult += evalError
        # print("    Erreur intermediaire", i, ":", evalError)

    print("Erreur moyenne sur l'ensemble de test :", round(100*testResult/numMbTest, 2), "%\n")
    print(str(datetime.datetime.now()))
    print("\n", "\n", "\n", "\n", "\n")

cntk.train.distributed.Communicator.finalize()`

**Comment by etienneFontaine35** [comment]
Here are the results for a 3 hidden layers with 1024 relu nodes on 2 GPUs :
- minibatch size = 64, training time = 53.7s, test error = 1.43%
- minibatch size = 128, training time = 29.5s, test error = 1.7%
- minibatch size = 256, training time = 19s, test error = 1.77%
- minibatch size = 512, training time = 13s, test error = 1.9%
- minibatch size = 1024, training time = 10.5s, test error = 2.26%
- minibatch size = 2048, training time = 9s, test error = 2.92%
- minibatch size = 4096, training time = 8s, test error = 4.44%

**Comment by etienneFontaine35** [comment]
Here are the results for a 3 hidden layers with 1024 relu nodes on **a single GPU** :
- minibatch size = 64, training time = 16.27s, test error = 1.41%
- minibatch size = 128, training time = 10.23s, test error = 1.63%
- minibatch size = 256, training time = 9s, test error = 1.75%
- minibatch size = 512, training time = 7.21s, test error = 1.69%
- minibatch size = 1024, training time = 6.68s, test error = 2.12%
- minibatch size = 2048, training time = 5s, test error = 4.47%
- minibatch size = 4096, training time = 5.24s, test error = 4.2%

**Comment by etienneFontaine35** [comment]
As you can see, single GPU running is always faster than multiple GPUs running with the same configuration (hopefully the error rates are the same).
I am also worrying about the losses which are the same for both GPUs at each step, here are the corresponding log :

2018-07-27 10:04:16.605497 

Input dimension : 784
Output dimension : 10
Hidden layers number : 3
Dimension layers : 1024
Epochs number : 20
Minibatch size : 1024 

Finished Epoch[1 of 20]: [Training] loss = 0.771140 * 60416, metric = 19.84% * 60416 1.232s (49039.0 samples/s);
Finished Epoch[1 of 20]: [Training] loss = 0.771140 * 60416, metric = 19.84% * 60416 1.726s (35003.5 samples/s);
Finished Epoch[2 of 20]: [Training] loss = 0.284031 * 60416, metric = 8.20% * 60416 0.710s (85093.0 samples/s);
Finished Epoch[2 of 20]: [Training] loss = 0.284031 * 60416, metric = 8.20% * 60416 0.710s (85093.0 samples/s);
Finished Epoch[3 of 20]: [Training] loss = 0.228922 * 59392, metric = 6.60% * 59392 0.412s (144155.3 samples/s);
Finished Epoch[3 of 20]: [Training] loss = 0.228922 * 59392, metric = 6.60% * 59392 0.412s (144155.3 samples/s);
Finished Epoch[4 of 20]: [Training] loss = 0.193977 * 60416, metric = 5.58% * 60416 0.482s (125344.4 samples/s);
Finished Epoch[4 of 20]: [Training] loss = 0.193977 * 60416, metric = 5.58% * 60416 0.482s (125344.4 samples/s);
Finished Epoch[5 of 20]: [Training] loss = 0.169181 * 59392, metric = 4.81% * 59392 0.406s (146285.7 samples/s);
Finished Epoch[5 of 20]: [Training] loss = 0.169181 * 59392, metric = 4.81% * 59392 0.406s (146285.7 samples/s);
Finished Epoch[6 of 20]: [Training] loss = 0.146011 * 60416, metric = 4.22% * 60416 0.485s (124569.1 samples/s);
Finished Epoch[6 of 20]: [Training] loss = 0.146011 * 60416, metric = 4.22% * 60416 0.485s (124569.1 samples/s);
Finished Epoch[7 of 20]: [Training] loss = 0.131846 * 60416, metric = 3.76% * 60416 0.412s (146640.8 samples/s);
Finished Epoch[7 of 20]: [Training] loss = 0.131846 * 60416, metric = 3.76% * 60416 0.412s (146640.8 samples/s);
Finished Epoch[8 of 20]: [Training] loss = 0.117131 * 59392, metric = 3.26% * 59392 0.300s (197973.3 samples/s);
Finished Epoch[8 of 20]: [Training] loss = 0.117131 * 59392, metric = 3.26% * 59392 0.300s (197973.3 samples/s);
Finished Epoch[9 of 20]: [Training] loss = 0.103977 * 60416, metric = 2.92% * 60416 0.240s (251733.3 samples/s);
Finished Epoch[9 of 20]: [Training] loss = 0.103977 * 60416, metric = 2.92% * 60416 0.240s (251733.3 samples/s);
Finished Epoch[10 of 20]: [Training] loss = 0.094978 * 59392, metric = 2.67% * 59392 0.414s (143458.9 samples/s);
Finished Epoch[10 of 20]: [Training] loss = 0.094978 * 59392, metric = 2.67% * 59392 0.414s (143458.9 samples/s);
Finished Epoch[11 of 20]: [Training] loss = 0.083245 * 60416, metric = 2.31% * 60416 0.433s (139528.9 samples/s);
Finished Epoch[11 of 20]: [Training] loss = 0.083245 * 60416, metric = 2.31% * 60416 0.433s (139528.9 samples/s);
Finished Epoch[12 of 20]: [Training] loss = 0.081351 * 60416, metric = 2.31% * 60416 0.605s (99861.2 samples/s);
Finished Epoch[12 of 20]: [Training] loss = 0.081351 * 60416, metric = 2.31% * 60416 0.605s (99861.2 samples/s);
Finished Epoch[13 of 20]: [Training] loss = 0.071513 * 59392, metric = 2.01% * 59392 0.438s (135598.2 samples/s);
Finished Epoch[13 of 20]: [Training] loss = 0.071513 * 59392, metric = 2.01% * 59392 0.438s (135598.2 samples/s);
Finished Epoch[14 of 20]: [Training] loss = 0.067580 * 60416, metric = 1.89% * 60416 0.598s (101030.1 samples/s);
Finished Epoch[14 of 20]: [Training] loss = 0.067580 * 60416, metric = 1.89% * 60416 0.598s (101030.1 samples/s);
Finished Epoch[15 of 20]: [Training] loss = 0.059236 * 59392, metric = 1.65% * 59392 0.504s (117841.3 samples/s);
Finished Epoch[15 of 20]: [Training] loss = 0.059236 * 59392, metric = 1.65% * 59392 0.504s (117841.3 samples/s);
Finished Epoch[16 of 20]: [Training] loss = 0.056777 * 60416, metric = 1.60% * 60416 0.604s (100026.5 samples/s);
Finished Epoch[16 of 20]: [Training] loss = 0.056777 * 60416, metric = 1.60% * 60416 0.604s (100026.5 samples/s);
Finished Epoch[17 of 20]: [Training] loss = 0.052748 * 60416, metric = 1.41% * 60416 0.411s (146997.6 samples/s);
Finished Epoch[17 of 20]: [Training] loss = 0.052748 * 60416, metric = 1.41% * 60416 0.411s (146997.6 samples/s);
Finished Epoch[18 of 20]: [Training] loss = 0.046556 * 59392, metric = 1.27% * 59392 0.593s (100155.1 samples/s);
Finished Epoch[18 of 20]: [Training] loss = 0.046556 * 59392, metric = 1.27% * 59392 0.593s (100155.1 samples/s);
Finished Epoch[19 of 20]: [Training] loss = 0.042679 * 60416, metric = 1.12% * 60416 0.602s (100358.8 samples/s);
Finished Epoch[19 of 20]: [Training] loss = 0.042679 * 60416, metric = 1.12% * 60416 0.602s (100358.8 samples/s);
Finished Epoch[20 of 20]: [Training] loss = 0.043155 * 59264, metric = 1.11% * 59264 0.612s (96836.6 samples/s);
Finished Epoch[20 of 20]: [Training] loss = 0.043155 * 59264, metric = 1.11% * 59264 0.612s (96836.6 samples/s);
Learning duration : 10.38 s
Learning duration : 10.86 s
Mean error on test set : 2.26 %

2018-07-27 10:04:27.834615

**Comment by etienneFontaine35** [comment]
Indeed aggregation duration is much longer than Forward  + Backward

CNTK Performance Profiler Summary Report

Time Stamp: 2018/08/01 15:54:34

Description................ ............Mean ..........StdDev .............Min .............Max ...........Count ...........Total

Main Thread

_Minibatch Iteration      :         4.408 ms        14.874 ms         2.925 ms       276.196 ms             2345     10337.365 ms
__Get Minibatch           :         1.170 ms        16.121 ms         0.028 ms       388.313 ms             2344      2742.374 ms
__Forward + Backward      :         0.970 ms         1.956 ms         0.764 ms        95.431 ms             2344      2274.470 ms
__Gradient Aggregation    :         3.133 ms        14.495 ms         1.779 ms       274.957 ms             2345      7347.257 ms
__Weight Update           :         0.206 ms         0.015 ms         0.001 ms         0.283 ms             2345       482.711 ms
__Post Processing         :         0.006 ms         0.020 ms         0.002 ms         0.819 ms             2345        13.435 ms

Data Reader

Prefetch Minibatch        :         1.615 ms        16.695 ms         0.304 ms       391.228 ms             2346      3789.288 ms

**Comment by etienneFontaine35** [comment]
One of the problem I know is that my GPUs are only using 8 PCIe lanes each, I plan to change my CPU so that both GPUs can benefit from having 16 PCIe lanes. Do you think it is worth it ? Because I am already applying gradient 1bit-quantization and there is only little difference (3.133ms gradient aggregation instead of 3.500 ms). I am using two 1080Ti, but I know Tesla GPUs have Nvlink, which is much faster than regular PCIe.
Do you know any other way to decrease communication time between GPUs ?

**Comment by KeDengMS** [comment]
If you are already using 1-bit quantization of gradient, the PCIe bandwidth should not matter much. I think you should try networks other than FF that have higher computation to weights ratio. IMO FF is the basic form of DNN, and CNN/RNN are forms of weight sharing on top of FF. By sharing weights, the model size would be reduced, thus aggregation cost takes smaller percentage of overall training.

Another suggestion is to try maximize your GPU memory usage before going distributed. A bigger minibatch size would increase computation occupancy with the same amount of gradient exchange per minibatch.

**Comment by etienneFontaine35** [comment]
Thank you for your response KeDengMS, I'll try the LSTM and figure out if it solves my problem.
